### PR TITLE
Implement workflow error event

### PR DESCRIPTION
**Pull request by hyzx86**
I created a workflow event that captures error messages when other workflow executions report errors

The code below is automatically generated when open  ` Catch Workflow Fault Event`  first time

image

### CODE REVIEW

**Comment by hyzx86** [comment]
@Skrypt  Can you review this PR for me?

File: src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/WorkflowFaultEventDisplayDriver.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/WorkflowFaultEventDisplayDriver.cs:15]: add new line between methods

File: src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/WorkflowFaultEventDisplayDriver.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Drivers/WorkflowFaultEventDisplayDriver.cs:18]: remove extra line

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:23]: Add a new line after. Consider using CTRL + k + d to format your code in Visual Studio.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:51]: Code Indentation.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:63]: new line before

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:80]: remove extra line

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:85]: remove last 2 lines

File: src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs:15]: same

**Review by Skrypt — CHANGES_REQUESTED** [review]
Fix broken Unit Tests.

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:51]: I don't think the above code needs to be indented here, otherwise there will be extra indentation when it is output in the editor

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:22]:
```csharp
    public WorkflowFaultEvent(
        IStringLocalizer<WorkflowFaultEvent> stringLocalizer,
        IWorkflowScriptEvaluator scriptEvaluator)
    {
            S = stringLocalizer;
            _scriptEvaluator = scriptEvaluator;
    }
```

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:25]: Remove extra lines

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:73]: New line before

**Review by hishamco — CHANGES_REQUESTED** [review]
Please fix all the formatting before review

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:23]: Add new line after each property

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:29]: Same here

**Review by hishamco — CHANGES_REQUESTED** [review]
Please format the PR properly

File: src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs:38]: Extra line

File: src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs:42]: Extra lines

File: src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs:26]: Extra line

File: src/OrchardCore.Modules/OrchardCore.Workflows/Startup.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Startup.cs:80]: Same here

**Review by hishamco — CHANGES_REQUESTED** [review]
Again, please check the format, before we get a real review ;)

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:22]: This is the effect of my visual studio formatting shortcuts....😕
image

**Comment by hyzx86** [comment]
@Skrypt @hishamco Hi guys，any question？

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:12]: Add space after the comma

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:13]: Same for the entire list

**Review by hishamco — CHANGES_REQUESTED** [review]
Just a little format to fix, after that I can approve. Also I need someone else from the team to approve

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:12]: Remove extra space after `nameof` in all the places

**Review by hishamco — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:12]: done

**Comment by hyzx86** [comment]
At present, this function finds an embarrassing problem,
I use this event to listen for all workflow exceptions
Once triggered, an email will be sent to my mailbox,
The actual situation is that the workflow is triggered, but the mail cannot be sent out
Is it because some objects have been released?

image

image

**Comment by hyzx86** [comment]
> The actual situation is that the workflow is triggered, but the mail cannot be sent out

Resolved ,  missing [Resume method](https://github.com/OrchardCMS/OrchardCore/pull/11539/commits/0759359ca5909312e204d711ab3b9a9eee98f5eb)

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:43]: Remove extra line

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:59]: Please check the format here

File: src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Handlers/DefaultWorkflowFaultHandler.cs:37]: Add new line before

File: src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs:420]: Add new line before

File: src/OrchardCore.Modules/OrchardCore.Workflows/Startup.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Startup.cs:76]: Remove the unnecessary comment

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:7]: Remove extra spaces

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:39]: Why you set the theme here, @Skrypt please your input here while your worked on the themes

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: Why there's too much hints?

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml:1]: Are we duplicate the titles in other thumbnails?

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml:1]: How about this one instead?

image

File: src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Events/WorkflowFaultEvent.cs:59]: These codes will generate default JavaScript  when creating events for users to refer to or modify directly.

If  format it, it will leave some blank space in the Monaco editor

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:39]: For darkmode switch

if remove that 
image


Refer to the usage of other positions, which should be

image

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: This Event contains many hidden contents (this is the case with many functions of OC at present).

I hope users can directly understand how to use this event through this page
When created, the editor renders the sample usage of all properties by default.
However, if the user modifies and saves the sample code, it may be lost.
In order to ensure that the reference documents can also be directly seen during modification, I have enumerated all the attributes below

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml:1]: I mean "Workflow Fault Event"

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: Can we use bulleted list as hint?

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: What about this?

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Thumbnail.cshtml:1]: updated to 
`Trigger and capture exception information when exception occurs in any other workflow.`

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: It would be nice if we add `hint` class to the list

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: updated

image

File: src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs:421]: It's strange that I have submitted the code and it is not updated here??

File: src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Views/Items/WorkflowFaultEvent.Fields.Edit.cshtml:8]: I think it's better

**Review by Skrypt — COMMENTED** [review]
Needs to be triaged on Thursday meeting

**Comment by hyzx86** [comment]
> Needs to be triaged on Thursday's meeting

What was the result of the discussion?
Has video playback been released?

**Comment by hishamco** [comment]
@hyzx86 it would be nice if you join the meeting. FYI there's no recording for the triage meeting

**Comment by hyzx86** [comment]
> @hyzx86 it would be nice if you join the meeting. FYI there's no recording for the triage meeting

I noticed that the meeting was held at 4 a. m. Chinese time..  😭

**Comment by hishamco** [comment]
Same here I'm suffering from Tuesday meeting due time zone difference :(

**Comment by Skrypt** [comment]
Seems like it did not get triaged or reviewed yet.

**Comment by Skrypt** [comment]
I'm waiting on @sebastienros review first so that we know if he wants to have that kind of implementation merged in OC.

**Review by sebastienros — APPROVED** [review]

**Pull request merged**
